src/flask_pyoidc/pyoidc_facade.py

```diff
@@ -143,13 +145,31 @@ def exchange_authorization_code(self, authorization_code: str, state: str):
         Union[AccessTokenResponse, TokenErrorResponse, None]
             The parsed token response, or None if no token request was performed.
         """
+        if not self._client.token_endpoint:
+            return None
+
         request_args = {
             'grant_type': 'authorization_code',
             'code': authorization_code,
             'redirect_uri': self._redirect_uri
         }
+        logger.debug('making token request: %s', request_args)
+        client_auth_method = self._client.registration_response.get('token_endpoint_auth_method',
+                                                                    'client_secret_basic')
+        auth_header = _ClientAuthentication(self._client.client_id, self._client.client_secret)(client_auth_method,
+                                                                                                request_args)
+        token_response = self._client.do_access_token_request(state=state,
+                                                              request_args=request_args,
+                                                              authn_method=client_auth_method,
+                                                              headers=auth_header,
+                                                              endpoint=self._client.token_endpoint
+                                                              )
+        logger.debug(f'received token response: {token_response}')
+        # Store token response in token_class instance. This is required when
+        # making refresh access token request.
+        self._client.token_class = Token(resp=token_response)
```

Comment:
Unfortunately this doesn't help, it only pushes the problem to another place.
The main problem (which I didn't explain clearly in my last comments) is that this will not work at all given an app where multiple users login.
Since this stores the token here after every token request it will always be overwritten with the last logged-in users tokens.
So imagine the following scenario:
1. User A logs into the app, the token response for User A is now stored.
1. User B logs into the app, the previous token response is replaced with token response for User B.
1. User A hits a scenario where their tokens needs refreshing -> the token response for User B will be used.

That is the reason why everything must be stateless - individual user tokens are stored in Flask session, which by definition is per-user.
Local variables like this are not, they are in a sense "global variables" used for all requests/users of the app.

Reply:
Fixed. The only reason I've to use Token class instance for refresh token request is because of this line of code in pyoidc which I'm not understanding why it has to be like this:

https://github.com/OpenIDC/pyoidc/blob/0d316fc0baca21b16164ef17fb275979d178f4b6/src/oic/oauth2/__init__.py#L497-L499

Why overwrite `refresh_token` in `request_args` when I'm already passing refresh_token in request_args. It's only because of this I'm initializing Token like this:

```python
self._client.do_access_token_refresh(request_args=request_args,
                                     authn_method=client_auth_method,
                                     token=Token(resp={'refresh_token': refresh_token}),
                                     endpoint=self._client.token_endpoint
                                     )
```

Reply:
Thanks - your new changes LGTM! 👍 

Yeah, I saw the reason for why you did it the original way, but I'm very happy you managed to work around it.
I think the intention of pyoidc, and its reason for overwriting the refresh token, is that it wants to manage all the tokens for you, i.e. it doesn't expect any `refresh_token` in the request_args (which in this use case becomes problematic).